[GTK][WPE] Stop using injected bundle messages for implementing resource loader API

carlosgcampos opened the pull request:
#### 051a04847864264224bafaea77cf12c5b185782f
<pre>
[GTK][WPE] Stop using injected bundle messages for implementing resource loader API
<a href="https://bugs.webkit.org/show_bug.cgi?id=245070">https://bugs.webkit.org/show_bug.cgi?id=245070</a>

Reviewed by Michael Catanzaro.

Use WebPageProxy messages instead of injected bundle messages for GLib
resources loader API implementation. Also deprecate
WebKitWebResource::data-received signal, since I haven&apos;t found any app
using it and it increases the IPC traffic.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitInjectedBundleClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebResource.cpp:
(webkit_web_resource_class_init):
(webkitWebResourceSentRequest):
(webkitWebResourceSetResponse):
(webkitWebResourceFinished):
(webkitWebResourceFailed):
(webkitWebResourceNotifyProgress): Deleted.
(webkitWebResourceFailedWithTLSErrors): Deleted.
(webkitWebResourceGetFrame): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebResourceLoadManager.cpp: Added.
(WebKit::WebKitWebResourceLoadManager::WebKitWebResourceLoadManager):
(WebKit::WebKitWebResourceLoadManager::~WebKitWebResourceLoadManager):
(WebKit::WebKitWebResourceLoadManager::didInitiateLoad):
(WebKit::WebKitWebResourceLoadManager::didSendRequest):
(WebKit::WebKitWebResourceLoadManager::didReceiveResponse):
(WebKit::WebKitWebResourceLoadManager::didFinishLoad):
* Source/WebKit/UIProcess/API/glib/WebKitWebResourceLoadManager.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebResourcePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::webResourceLoadManager):
(webkitWebViewConstructed):
(webkitWebViewLoadChanged):
(webkitWebViewGetWebResourceLoadManager):
(webkitWebViewResourceLoadStarted):
(webkitWebViewGetLoadingWebResource): Deleted.
(webkitWebViewRemoveLoadingWebResource): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::webResourceLoadManager):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/glib/WebPageProxyGLib.cpp:
(WebKit::WebPageProxy::didInitiateLoadForResource):
(WebKit::WebPageProxy::didSendRequestForResource):
(WebKit::WebPageProxy::didReceiveResponseForResource):
(WebKit::WebPageProxy::didFinishLoadForResource):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::assignIdentifierToInitialRequest):
(WebKit::WebFrameLoaderClient::dispatchWillSendRequest):
(WebKit::WebFrameLoaderClient::dispatchDidReceiveResponse):
(WebKit::WebFrameLoaderClient::dispatchDidFinishLoading):
(WebKit::WebFrameLoaderClient::dispatchDidFailLoading):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestResources.cpp:
(testWebResourceLoading):
(testWebResourceSendRequest):

Canonical link: <a href="https://commits.webkit.org/256388@main">https://commits.webkit.org/256388@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/953f8dc7c2b8b01dc23992b645a67502b536fcaa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95413 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4686 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/28461 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104995 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/165257 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99400 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4701 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33413 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87779 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100868 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/101077 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/3431 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/82026 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30504 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/85326 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/87238 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/73344 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/39204 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18792 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36907 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/20088 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4416 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40884 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/42752 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42873 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/39333 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -494,6 +493,11 @@ void WebKitWebViewClient::didReceiveUserMessage(WKWPE::View&, UserMessage&& mess
     webkitWebViewDidReceiveUserMessage(m_webView, WTFMove(message), WTFMove(completionHandler));
 }
+
+WebKitWebResourceLoadManager* WebKitWebViewClient::webResourceLoadManager()
```
It should return `WebKitWebResourceLoadManager&` IMO, because once the object is constructed it should never be `nullptr`. This will make it easier to keep track of what is and isn't nullable. You can add an assert to check that it's not null before dereferencing here.

(I think it's worth doing this for GObject types too, since it's a nice safety improvement. But we've intentionally avoided that so far. This one is not a GObject, though. ;)

carlosgcampos replied:
No, it's nullptr when GLib API is not used (WTR)

mcatanzaro commented on Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
```diff
@@ -632,4 +632,9 @@ WebCore::Color PageClientImpl::accentColor()
 
+WebKitWebResourceLoadManager* PageClientImpl::webResourceLoadManager()
+{
+    return WEBKIT_IS_WEB_VIEW(m_viewWidget) ? webkitWebViewGetWebResourceLoadManager(WEBKIT_WEB_VIEW(m_viewWidget)) : nullptr;
```
Can you remind me how we can get a `PageClientImpl` with a `m_viewWidget` that is not a `WebKitWebView`? It looks like it's only constructed from `WebKitWebViewBase`, and `PageClientImpl` does assume pervasively that it is of that type. Is it possible to have a `WebKitWebViewBase` that is not a `WebKitWebView`? Well, it is a public type, so I suppose an application could create one directly... but can it happen normally? If it's not normal, then I bet we can simplify things by removing more nullability.

carlosgcampos replied:
Yes, when using the C API you have a `WebKitWebViewBase` that is not a `WebKitWebView`, that's why we have base class. In the case of wpe the view client is not implemented for WTR.

carlosgcampos replied:
The web inspector also uses WebKitWebViewBase only

carlosgcampos committed:
[GTK][WPE] Stop using injected bundle messages for implementing resource loader API
https://bugs.webkit.org/show_bug.cgi?id=245070

Reviewed by Michael Catanzaro.

Use WebPageProxy messages instead of injected bundle messages for GLib
resources loader API implementation. Also deprecate
WebKitWebResource::data-received signal, since I haven't found any app
using it and it increases the IPC traffic.

* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitInjectedBundleClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebResource.cpp:
(webkit_web_resource_class_init):
(webkitWebResourceSentRequest):
(webkitWebResourceSetResponse):
(webkitWebResourceFinished):
(webkitWebResourceFailed):
(webkitWebResourceNotifyProgress): Deleted.
(webkitWebResourceFailedWithTLSErrors): Deleted.
(webkitWebResourceGetFrame): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebResourceLoadManager.cpp: Added.
(WebKit::WebKitWebResourceLoadManager::WebKitWebResourceLoadManager):
(WebKit::WebKitWebResourceLoadManager::~WebKitWebResourceLoadManager):
(WebKit::WebKitWebResourceLoadManager::didInitiateLoad):
(WebKit::WebKitWebResourceLoadManager::didSendRequest):
(WebKit::WebKitWebResourceLoadManager::didReceiveResponse):
(WebKit::WebKitWebResourceLoadManager::didFinishLoad):
* Source/WebKit/UIProcess/API/glib/WebKitWebResourceLoadManager.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitWebResourcePrivate.h:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(WebKitWebViewClient::webResourceLoadManager):
(webkitWebViewConstructed):
(webkitWebViewLoadChanged):
(webkitWebViewGetWebResourceLoadManager):
(webkitWebViewResourceLoadStarted):
(webkitWebViewGetLoadingWebResource): Deleted.
(webkitWebViewRemoveLoadingWebResource): Deleted.
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.cpp:
(WebKit::PageClientImpl::webResourceLoadManager):
* Source/WebKit/UIProcess/API/gtk/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/APIViewClient.h:
(API::ViewClient::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.cpp:
(WebKit::PageClientImpl::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/PageClientImpl.h:
* Source/WebKit/UIProcess/API/wpe/WPEView.cpp:
(WKWPE::View::webResourceLoadManager):
* Source/WebKit/UIProcess/API/wpe/WPEView.h:
* Source/WebKit/UIProcess/API/wpe/WebKitWebViewClient.h:
* Source/WebKit/UIProcess/PageClient.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/UIProcess/glib/WebPageProxyGLib.cpp:
(WebKit::WebPageProxy::didInitiateLoadForResource):
(WebKit::WebPageProxy::didSendRequestForResource):
(WebKit::WebPageProxy::didReceiveResponseForResource):
(WebKit::WebPageProxy::didFinishLoadForResource):
* Source/WebKit/WebProcess/InjectedBundle/API/glib/WebKitWebPage.cpp:
* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::assignIdentifierToInitialRequest):
(WebKit::WebFrameLoaderClient::dispatchWillSendRequest):
(WebKit::WebFrameLoaderClient::dispatchDidReceiveResponse):
(WebKit::WebFrameLoaderClient::dispatchDidFinishLoading):
(WebKit::WebFrameLoaderClient::dispatchDidFailLoading):
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestResources.cpp:
(testWebResourceLoading):
(testWebResourceSendRequest):

Canonical link: https://commits.webkit.org/256388@main
